tests/test_admin.py

```diff
@@ -248,6 +263,147 @@ def test_advanced_settings_link(self):
+        PageVersionFactory(
+            content__page__node__depth=1, content__page__is_home=1, state=PUBLISHED
+        )
+        version_2 = PageVersionFactory(content__page__node__depth=1, state=PUBLISHED)
+        to_be_homepage = version_2.content
+        with self.login_user_context(self.get_superuser()), patch(
+            "cms.models.query.PageQuerySet.has_apphooks", return_value=True
+        ), patch(
+            "djangocms_pageadmin.admin.set_restart_trigger", return_value=True
+        ) as mock_handler:
+            self.client.post(
+                self.get_admin_url(PageContent, "set_home_content", to_be_homepage.pk)
+            )
+            mock_handler.assert_called()
+
+    def test_when_old_homepage_tree_has_no_apphooks_shouldnt_trigger_signal(self):
+        PageVersionFactory(
+            content__page__node__depth=1, content__page__is_home=1, state=PUBLISHED
+        )
+        version_2 = PageVersionFactory(content__page__node__depth=1, state=PUBLISHED)
+        to_be_homepage = version_2.content
+        with self.login_user_context(self.get_superuser()), patch(
+            "cms.models.query.PageQuerySet.has_apphooks", return_value=False
+        ) as mock, patch(
+            "djangocms_pageadmin.admin.set_restart_trigger"
+        ) as mock_handler:
+            self.client.post(
+                self.get_admin_url(PageContent, "set_home_content", to_be_homepage.pk)
+            )
+
+            mock_handler.assert_not_called()
+
+    def test_when_old_home_tree_is_none_should_not_trigger_signal(self):
+        version_1 = PageVersionFactory(
+            content__page__node__depth=1, content__page__is_home=1, state=PUBLISHED
+        )
+        homepage = version_1.content
+        with self.login_user_context(self.get_superuser()), patch(
+            "cms.models.query.PageQuerySet.has_apphooks", return_value=False
+        ) as mock, patch("cms.signals.apphook.set_restart_trigger") as mock_handler:
+            self.client.post(
+                self.get_admin_url(PageContent, "set_home_content", homepage.pk)
+            )
+
+            mock_handler.assert_not_called()
+
+
+class SetHomeViewTransactionTestCase(TransactionTestCase):
+    def setUp(self):
+        self.user = UserFactory(is_staff=True, is_superuser=True)
+        # Login
+        self.client.force_login(self.user)
+
+    def test_set_home_is_wrapped_in_db_transaction(self):
+        class FakeError(Exception):
+            pass
+
+        version_1 = PageVersionFactory(content__page__node__depth=1, state=PUBLISHED)
+        homepage = version_1.content
+
+        # Asserting to make sure page is not already set as home
+        self.assertFalse(homepage.page.is_home)
+
+        # Patching app_hooks which has been called after setting home on view so transaction
+        # should roll back in case of error
+        with patch("cms.models.query.PageQuerySet.has_apphooks", side_effect=FakeError):
```

Comment:
Just checking, but does this patch definitely work? (i.e. make that throw that error) As in if you comment out the transaction.atomic() decorator this test fails? It's just that this is not really the recommended way to patch (usually the mocking docs suggest patching the place to which something is imported, which in this case is not really possible) and from experience, this can backfire (i.e. the mock just doesn't happen where you expect it to etc.).

Reply:
@monikasulik has_apphooks haven't been imported and it was a method of page_tree (new_page_tree, old_page_tree in view) hence patch working. I tested manually by commenting tranaction.atomic() line and test is breaking.